usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2974,6 +2974,304 @@ static CK_BBOOL ccatok_pkey_option_disabled(STDLL_TokData_t *tokdata)
+    if (ret != CKR_OK) {
+        TRACE_ERROR("pkey_update_and_save failed with rc=0x%lx\n", ret);
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    if (pkey_attr != NULL)
+        free(pkey_attr);
+
+    return ret;
+}
+
+/**
+ * Returns true if the session is ok for creating protected keys, false
+ * otherwise. The session must be read/write for token objects, and not public
+ * nor SO for private objects.
+ */
+static CK_BBOOL ccatok_pkey_session_ok_for_obj(SESSION *session,
+                                               OBJECT *key_obj)
+{
+    if (object_is_token_object(key_obj) &&
+        (session->session_info.flags & CKF_RW_SESSION) == 0)
+        return CK_FALSE;
+
+    if (object_is_private(key_obj)) {
+        switch (session->session_info.state) {
+        case CKS_RO_PUBLIC_SESSION:
+        case CKS_RW_PUBLIC_SESSION:
+        case CKS_RW_SO_FUNCTIONS:
+            return CK_FALSE;
+        default:
+            break;
+        }
+    }
+
+    return CK_TRUE;
+}
+
+/**
+ * Checks if the preconditions for using the related protected key of
+ * the given secure key object are met. The caller of this routine must
+ * have a READ_LOCK on the key object.
+ *
+ * The routine internally creates a protected key and adds it to the key_obj,
+ * if the machine supports pkeys, the key is eligible for pkey support, does
+ * not already have a valid pkey, and other conditions, like r/w session, are
+ * fulfilled. As adding a protected key to the key_obj involves unlocking and
+ * re-locking, the key blob, or any other attribute of the key, that was
+ * retrieved via h_opaque_2_blob before calling this function might be no more
+ * valid in a parallel environment.
+ *
+ * Therefore, the following return codes tell the calling function how to
+ * proceed:
+ *
+ * @return CKR_OK:
+ *            a protected key was possibly created successfully and everything
+ *            is fine to use pkey support. In this case the protected key
+ *            shall be used, but a previously obtained key blob or other attr
+ *            might be invalid, because of a possible unlock/re-lock of the
+ *            key_obj.
+ *
+ *         CKR_FUNCTION_NOT_SUPPORTED:
+ *            The system, session or key do not allow to use pkey support, but
+ *            no attempt was made to create a protected key. So the key blob,
+ *            or any other attr, is still valid and a fallback into the ep11
+ *            path is ok.
+ *
+ *         all others:
+ *            An internal error occurred and it was possibly attempted to create
+ *            a protected key for the object. In this case, the key blob, or
+ *            any other attr, might be no longer valid in a parallel environment
+ *            and the ep11 fallback is not possible anymore. The calling
+ *            function shall return with an error in this case.
+ */
+CK_RV ccatok_pkey_check(STDLL_TokData_t *tokdata, SESSION *session,
```

Comment:
I guess this function can also be static. You are not using it anywhere outside of cca_specific.c.